File: Pythoncode/33.搜索旋转排序数组.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if len(nums) == 0:
            return -1
        
        start = 0
        end = len(nums) - 1

        if target >= nums[0]: # target on the left
            while start <= end:
                mid = (start + end) // 2
                if nums[mid] < nums[start] or target < nums[mid]:
                    end = mid - 1
                elif target > nums[mid]:
                    start = mid + 1
                else:
                    return mid

        else: #target on the right
            while start <= end:
                mid = (start + end) // 2
                if nums[mid] > nums[end] or target > nums[mid]:
                    start = mid + 1
                elif target < nums[mid] :
                    end = mid - 1 
                else:
                    return mid
        
        return -1

File: Cppcode/33.搜索旋转排序数组.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        return helper(nums, target, 0, nums.size()-1);
    }

    int helper(vector<int>& nums, int target, int head, int tail)
    {
        if (tail < head) {
            return -1;
        }
        auto mid = (head + tail) / 2;
        if (nums[head] <= nums[tail])
        {
            if (target == nums[mid])
            {
                return mid;
            }
            else if (target < nums[mid])
            {
                return helper(nums, target, head, mid-1);
            }
            else
            {
                return helper(nums, target, mid+1, tail);
            }
        }
        else
        {
            if (target == nums[mid])
            {
                return mid;
            }
            if (nums[mid] <= nums[tail])
            {
                if (target < nums[mid])
                {
                    return helper(nums, target, head, mid-1);
                }
                else if (target > nums[mid] && target <= nums[tail]) {
                    return helper(nums, target, mid+1, tail);
                }
                else
                {
                    return helper(nums, target, head, mid-1);
                }
            }
            else // nums[mid] >= nums[head]
            {
                if (target > nums[mid])
                {
                    return helper(nums, target, mid+1, tail);
                }
                else if (target < nums[mid] && target >= nums[head])
                {
                    return helper(nums, target, head, mid-1);
                }
                else
                {
                    return helper(nums, target, mid+1, tail);
                }
            }
        }
    }
};
